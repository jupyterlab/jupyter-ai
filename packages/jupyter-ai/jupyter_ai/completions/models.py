from typing import List, Literal, Optional

from langchain.pydantic_v1 import BaseModel


class InlineCompletionRequest(BaseModel):
    """Message send by client to request inline completions.

    Prefix/suffix implementation is used to avoid the need for synchronising
    the notebook state at every key press (subject to change in future)."""

    # unique message ID generated by the client used to identify replies and
    # to easily discard replies for older requests
    number: int
    # prefix should include full text of the current cell preceding the cursor
    prefix: str
    # suffix should include full text of the current cell preceding the cursor
    suffix: str
    # media type for the current language, e.g. `text/x-python`
    mime: str
    # whether to stream the response (if supported by the model)
    stream: bool
    # path to the notebook of file for which the completions are generated
    path: Optional[str]
    # language inferred from the document mime type (if possible)
    language: Optional[str]
    # identifier of the cell for which the completions are generated if in a notebook
    # previous cells and following cells can be used to learn the wider context
    cell_id: Optional[str]


class InlineCompletionItem(BaseModel):
    """The inline completion suggestion to be displayed on the frontend.

    See JuptyerLab `InlineCompletionItem` documentation for the details.
    """

    insertText: str
    filterText: Optional[str]
    isIncomplete: Optional[bool]
    token: Optional[str]


class CompletionError(BaseModel):
    type: str
    traceback: str


class InlineCompletionList(BaseModel):
    """Reflection of JupyterLab's `IInlineCompletionList`."""

    items: List[InlineCompletionItem]


class InlineCompletionReply(BaseModel):
    """Message sent from model to client with the infill suggestions"""

    list: InlineCompletionList
    # number of request for which we are replying
    reply_to: int
    error: Optional[CompletionError]


class InlineCompletionStreamChunk(BaseModel):
    """Message sent from model to client with the infill suggestions"""

    type: Literal["stream"] = "stream"
    response: InlineCompletionItem
    reply_to: int
    done: bool
    error: Optional[CompletionError]
