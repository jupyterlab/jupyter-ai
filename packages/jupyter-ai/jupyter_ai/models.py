from pydantic import BaseModel
from typing import Dict, List

class PromptRequest(BaseModel):
    task_id: str
    engine_id: str
    prompt_variables: Dict[str, str]

class ListEnginesEntry(BaseModel):
    id: str
    name: str

class ListTasksEntry(BaseModel):
    id: str
    name: str

class ListTasksResponse(BaseModel):
    tasks: List[ListTasksEntry]

class DescribeTaskResponse(BaseModel):
    name: str
    insertion_mode: str
    prompt_template: str
    engines: List[ListEnginesEntry]
