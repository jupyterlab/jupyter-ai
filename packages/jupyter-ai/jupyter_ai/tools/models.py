import re
from typing import Callable, Optional

from pydantic import BaseModel, ConfigDict, Field, model_validator


def get_doc_description(func) -> str:
    """Extract the first paragraph from a function's docstring.

    Args:
        func: The function to extract documentation from.

    Returns:
        The first paragraph of the function's docstring, cleaned up and
        with whitespace normalized. Returns empty string if no docstring exists.

    Example:
        >>> def sample_func():
        ...     '''This is the first paragraph.
        ...
        ...     This is the second paragraph.'''
        ...     pass
        >>> get_doc_description(sample_func)
        'This is the first paragraph.'
    """
    if not func.__doc__:
        return ""

    # Split docstring into paragraphs
    paragraphs = re.split(r"\n\s*\n", func.__doc__.strip())

    # Return the first paragraph, cleaned up
    if paragraphs:
        return re.sub(r"\s+", " ", paragraphs[0].strip())

    return ""


class Tool(BaseModel):
    """Represents a tool with callable function and capability metadata.

    A Tool wraps a callable function with metadata including name, description,
    and flags that indicate what operations the tool can perform.

    Attributes:
        callable: The function that implements the tool's functionality.
        name: The tool's name. Auto-generated from callable if not provided.
        description: Tool description. Auto-extracted from callable docstring if not provided.
        read: Whether the tool can read data.
        write: Whether the tool can write data.
        execute: Whether the tool can execute operations.
        delete: Whether the tool can delete data.

    Example:
        >>> async def greet(name: str):
        ...     '''Say hello to someone.'''
        ...     return f"Hello, {name}!"
        >>> tool = Tool(callable=greet, read=True)
        >>> tool.name
        'greet'
        >>> tool.description
        'Say hello to someone.'
        >>> tool.read
        True
    """

    callable: Callable = Field(exclude=True)
    name: Optional[str] = None
    description: Optional[str] = None
    read: bool = False
    write: bool = False
    execute: bool = False
    delete: bool = False

    @model_validator(mode="after")
    def set_name_description(self):
        if not self.name:
            if hasattr(self.callable, "__name__") and self.callable.__name__:
                self.name = self.callable.__name__
            else:
                raise ValueError("Unable to extract name from callable")

        if not self.description:
            self.description = get_doc_description(self.callable)

        return self

    def __eq__(self, other):
        if not isinstance(other, Tool):
            return False
        return (
            self.name == other.name
            and self.description == other.description
            and self.read == other.read
            and self.write == other.write
            and self.delete == other.delete
            and self.execute == other.execute
        )

    def __hash__(self):
        return hash(id(self.callable))


class Toolkit(BaseModel):
    """A collection of tools with capability-based filtering.

    A Toolkit groups related tools together and provides methods to add tools
    and find tools based on their capability flags.

    Attributes:
        name: The toolkit's name.
        description: Optional description of the toolkit.
        tools: The set of tools in this toolkit.

    Example:
        >>> def read_file(path: str):
        ...     '''Read a file from disk.'''
        ...     with open(path) as f:
        ...         return f.read()
        >>> def write_file(path: str, content: str):
        ...     '''Write content to a file.'''
        ...     with open(path, 'w') as f:
        ...         f.write(content)
        >>> toolkit = Toolkit(name="FileTools", description="File operations")
        >>> read_tool = Tool(callable=read_file, read=True)
        >>> write_tool = Tool(callable=write_file, write=True)
        >>> toolkit.add_tool(read_tool)
        >>> toolkit.add_tool(write_tool)
        >>> len(toolkit.tools)
        2
        >>> read_tools = toolkit.find_tools(read=True)
        >>> len(read_tools)
        1
    """

    name: str
    description: Optional[str] = None
    tools: set = Field(default_factory=set)
    model_config = ConfigDict(arbitrary_types_allowed=True)

    def add_tool(self, tool: Tool):
        """Add a tool to this toolkit.

        Args:
            tool: The tool to add to the toolkit.
        """
        self.tools.add(tool)

    def get_tools(
        self,
        read: Optional[bool] = None,
        write: Optional[bool] = None,
        execute: Optional[bool] = None,
        delete: Optional[bool] = None,
    ) -> set[Tool]:
        """Find tools in this toolkit based on capability filters.

        Returns tools that match all of the specified capability criteria.
        If a capability filter is None, it won't be considered in filtering.
        If all capability filters are None, returns all tools in the toolkit.

        Args:
            read: Whether the tool can read data. None means don't filter on this capability.
            write: Whether the tool can write data. None means don't filter on this capability.
            execute: Whether the tool can execute operations. None means don't filter on this capability.
            delete: Whether the tool can delete data. None means don't filter on this capability.

        Returns:
            A set containing tools that match the specified capability criteria.

        Example:
            >>> toolkit = Toolkit(name="TestToolkit")
            >>> read_tool = Tool(callable=lambda: None, name="reader", read=True)
            >>> write_tool = Tool(callable=lambda: None, name="writer", write=True)
            >>> write_execute_tool = Tool(callable=lambda: None, name="writer_executor", write=True, execute=True)
            >>> toolkit.add_tool(read_tool)
            >>> toolkit.add_tool(write_tool)
            >>> toolkit.add_tool(write_execute_tool)
            >>> read_tools = toolkit.get_tools(read=True)
            >>> len(read_tools)
            1
            >>> all_tools = toolkit.get_tools()
            >>> len(all_tools)
            3
            >>> write_tools = toolkit.get_tools(write=True)
            >>> len(write_tools)
            2
            >>> write_and_execute_tools = toolkit.get_tools(write=True, execute=True)
            >>> len(write_and_execute_tools)
            1
        """
        toolset = set()

        for tool in self.tools:
            # Tool should match all non-None filter criteria
            if (
                (read is None or read == tool.read)
                and (write is None or write == tool.write)
                and (execute is None or execute == tool.execute)
                and (delete is None or delete == tool.delete)
            ):
                toolset.add(tool)

        return toolset
