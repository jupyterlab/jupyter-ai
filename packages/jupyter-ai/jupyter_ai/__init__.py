from ._version import __version__
from .extension import AiExtension
from .magics import AiMagics

# imports to expose entry points. DO NOT REMOVE.
from .engine import GPT3ModelEngine
from .tasks import tasks
from .providers import (
    AI21Provider,
    AnthropicProvider,
    CohereProvider,
    HfHubProvider,
    OpenAIProvider,
    ChatOpenAIProvider,
    SmEndpointProvider
)

# imports to expose types to other AI modules. DO NOT REMOVE.
from .providers import BaseProvider
from .tasks import DefaultTaskDefinition

def _jupyter_labextension_paths():
    return [{
        "src": "labextension",
        "dest": "@jupyter-ai/core"
    }]

def _jupyter_server_extension_points():
    return [{
        "module": "jupyter_ai",
        "app": AiExtension
    }]

def load_ipython_extension(ipython):
    ipython.register_magics(AiMagics)
