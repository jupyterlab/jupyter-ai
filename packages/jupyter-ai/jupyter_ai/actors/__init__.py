"""Actor classes that process incoming chat messages"""
