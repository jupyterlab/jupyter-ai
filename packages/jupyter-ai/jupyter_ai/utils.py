import openai


class LLMErrorUtilBase:
    @staticmethod
    def is_api_key_exc(e: Exception):
        raise NotImplementedError("Should be implemented by subclasses.")


class OpenAIErrorUtil(LLMErrorUtilBase):
    @staticmethod
    def is_api_key_exc(e: Exception):
        if isinstance(e, openai.error.AuthenticationError):
            error_details = e.json_body.get("error", {})
            return error_details.get("code") == "invalid_api_key"
        return False


class AI21ErrorUtility:
    @staticmethod
    def is_api_key_exc(e: Exception):
        if isinstance(e, ValueError):
            # Check if the exception message contains "status code 401"
            return "status code 401" in str(e)
        return False
