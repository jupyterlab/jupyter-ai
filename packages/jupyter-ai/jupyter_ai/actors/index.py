import time
from uuid import uuid4
from jupyter_ai.models import AgentChatMessage, HumanChatMessage
from langchain import FAISS
import ray
import os

from jupyter_ai.actors.base import BaseActor, Logger
from jupyter_ai_magics.providers import ChatOpenAINewProvider

from ray.util.queue import Queue
from jupyter_core.paths import jupyter_data_dir
from langchain.embeddings.openai import OpenAIEmbeddings
from langchain.document_loaders import TextLoader
from langchain.text_splitter import RecursiveCharacterTextSplitter
from jupyter_ai.document_loaders.directory import DirectoryLoader


@ray.remote
class DocumentIndexActor(BaseActor):
    def __init__(self, reply_queue: Queue, root_dir: str, log: Logger):
        super().__init__(log=log, reply_queue=reply_queue)
        self.root_dir = root_dir
        self.index_save_dir = os.path.join(jupyter_data_dir(), '.jupyter_ai_indexes')

        if ChatOpenAINewProvider.auth_strategy.name not in os.environ:
            return
        
        if not os.path.exists(self.index_save_dir):
            os.makedirs(self.index_save_dir)

        embeddings = OpenAIEmbeddings()
        try:
            self.index = FAISS.load_local(self.index_save_dir, embeddings)
        except Exception as e:
            self.index = FAISS.from_texts(["This is just starter text for the index"], embeddings)

    def get_index(self):
        return self.index

    def process_message(self, message: HumanChatMessage):
        dir_path = message.body.split(' ', 1)[-1]
        load_path = os.path.join(self.root_dir, dir_path)
        loader = DirectoryLoader(
            load_path, 
            glob=['*.txt', '*.text', '*.md', '*.py', '*.ipynb', '*.R'],
            loader_cls=TextLoader
        )
        documents = loader.load_and_split(
            text_splitter=RecursiveCharacterTextSplitter(chunk_size=1000, chunk_overlap=20)
        )
        self.index.add_documents(documents)
        self.index.save_local(self.index_save_dir)

        response = f"""ðŸŽ‰ I have indexed documents at **{dir_path}** and ready to answer questions. 
        You can ask questions from these docs by prefixing your message with **/fs**."""
        agent_message = AgentChatMessage(
                id=uuid4().hex,
                time=time.time(),
                body=response,
                reply_to=message.id
            )
        self.reply_queue.put(agent_message)