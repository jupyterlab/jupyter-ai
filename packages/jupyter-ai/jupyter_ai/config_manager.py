import json
import logging
import os
from typing import Any, Dict, Union

from jupyter_ai.models import GlobalConfig
from jupyter_ai_magics.utils import (
    AnyProvider,
    EmProvidersDict,
    LmProvidersDict,
    get_em_provider,
    get_lm_provider,
)
from jupyter_core.paths import jupyter_data_dir

Logger = Union[logging.Logger, logging.LoggerAdapter]


class ConfigManager:
    """Provides model and embedding provider id along
    with the credentials to authenticate providers.
    """

    def __init__(
        self, log: Logger, lm_providers: LmProvidersDict, em_providers: EmProvidersDict
    ):
        self.log = log
        self.save_dir = os.path.join(jupyter_data_dir(), "jupyter_ai")
        self.save_path = os.path.join(self.save_dir, "config.json")
        self.config = None
        self.lm_providers = lm_providers
        self.em_providers = em_providers
        self.lm_provider = None
        self.lm_provider_params = {}
        self.em_provider = None
        self.em_provider_params = {}
        self._load()

    def update(self, config: GlobalConfig, save_to_disk: bool = True):
        self._update_lm_provider(config)
        self._update_em_provider(config)
        if save_to_disk:
            self._save(config)
        self.config = config

    def get_config(self):
        return self.config

    def get_lm_provider(self):
        return self.lm_provider

    def get_lm_provider_params(self):
        return self.lm_provider_params

    def get_em_provider(self):
        return self.em_provider

    def get_em_provider_params(self):
        return self.em_provider_params

    def _authenticate_provider(
        self,
        provider: AnyProvider,
        provider_params: Dict[str, Any],
        config: GlobalConfig,
    ):
        auth_strategy = provider.auth_strategy
        if auth_strategy and auth_strategy.type == "env":
            api_keys = config.api_keys
            name = auth_strategy.name
            if name not in api_keys:
                raise ValueError(
                    f"Missing value for '{auth_strategy.name}' in the config."
                )
            provider_params[name.lower()] = api_keys[name]

    def _update_lm_provider(self, config: GlobalConfig):
        model_id = config.model_provider_id

        if not model_id:
            self.lm_provider = None
            self.lm_provider_params = None
            return

        local_model_id, provider = get_lm_provider(model_id, self.lm_providers)

        if not provider:
            raise ValueError(f"No provider and model found with '{model_id}'")

        fields = config.fields.get(model_id, {})
        provider_params = {"model_id": local_model_id, **fields}

        self._authenticate_provider(provider, provider_params, config)
        self.lm_provider = provider
        self.lm_provider_params = provider_params

    def _update_em_provider(self, config: GlobalConfig):
        model_id = config.embeddings_provider_id

        if not model_id:
            self.em_provider = None
            self.em_provider_params
            return

        local_model_id, provider = get_em_provider(model_id, self.em_providers)

        if not provider:
            raise ValueError(f"No provider and model found with '{model_id}'")

        provider_params = {"model_id": local_model_id}

        self._authenticate_provider(provider, provider_params, config)
        self.em_provider = provider
        self.em_provider_params = provider_params

    def _save(self, config: GlobalConfig):
        if not os.path.exists(self.save_dir):
            os.makedirs(self.save_dir)

        with open(self.save_path, "w") as f:
            f.write(config.json())

    def _load(self):
        if os.path.exists(self.save_path):
            with open(self.save_path, encoding="utf-8") as f:
                config = GlobalConfig(**json.loads(f.read()))
                self.update(config, False)
            return

        # otherwise, create a new empty config file
        self.update(GlobalConfig(), True)
