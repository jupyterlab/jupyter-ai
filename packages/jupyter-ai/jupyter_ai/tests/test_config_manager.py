import json
import logging
import os
from unittest.mock import mock_open, patch

import pytest
from jupyter_ai.config_manager import (
    AuthError,
    ConfigManager,
    KeyInUseError,
    WriteConflictError,
)
from jupyter_ai.models import DescribeConfigResponse, GlobalConfig, UpdateConfigRequest
from jupyter_ai_magics.utils import get_em_providers, get_lm_providers
from langchain.pydantic_v1 import ValidationError


@pytest.fixture
def config_path(jp_data_dir):
    return str(jp_data_dir / "config.json")


@pytest.fixture
def schema_path(jp_data_dir):
    return str(jp_data_dir / "config_schema.json")


@pytest.fixture
def common_cm_kwargs(config_path, schema_path):
    """Kwargs that are commonly used when initializing the CM."""
    log = logging.getLogger()
    lm_providers = get_lm_providers()
    em_providers = get_em_providers()
    return {
        "log": log,
        "lm_providers": lm_providers,
        "em_providers": em_providers,
        "config_path": config_path,
        "schema_path": schema_path,
        "allowed_providers": None,
        "blocked_providers": None,
        "allowed_models": None,
        "blocked_models": None,
    }


@pytest.fixture
def cm(common_cm_kwargs):
    """The default ConfigManager instance, with an empty config and config schema."""
    return ConfigManager(**common_cm_kwargs)


@pytest.fixture
def cm_with_blocklists(common_cm_kwargs):
    kwargs = {
        **common_cm_kwargs,
        "blocked_providers": ["ai21"],
        "blocked_models": ["cohere:medium"],
    }
    return ConfigManager(**kwargs)


@pytest.fixture
def cm_with_allowlists(common_cm_kwargs):
    kwargs = {
        **common_cm_kwargs,
        "allowed_providers": ["ai21"],
        "allowed_models": ["cohere:medium"],
    }
    return ConfigManager(**kwargs)


@pytest.fixture(autouse=True)
def reset(config_path, schema_path):
    """Fixture that deletes the config and config schema after each test."""
    yield
    try:
        os.remove(config_path)
    except OSError:
        pass
    try:
        os.remove(schema_path)
    except OSError:
        pass


@pytest.fixture
def config_with_bad_provider_ids(tmp_path):
    """Fixture that creates a `config.json` with `model_provider_id` and  `embeddings_provider_id` values that would not associate with models. File is created in `tmp_path` folder. Function returns path to the file."""
    config_data = {
        "model_provider_id:": "foo:bar",
        "embeddings_provider_id": "buzz:fizz",
        "api_keys": {},
        "send_with_shift_enter": False,
        "fields": {},
    }
    config_path = tmp_path / "config.json"
    with open(config_path, "w") as file:
        json.dump(config_data, file)
    return str(config_path)


@pytest.fixture
def cm_with_bad_provider_ids(common_cm_kwargs, config_with_bad_provider_ids):
    """Config manager instance created with `config_path` set to mocked `config.json` with `model_provider_id` and `embeddings_provider_id` values that would not associate with models."""
    common_cm_kwargs["config_path"] = config_with_bad_provider_ids
    return ConfigManager(**common_cm_kwargs)


def configure_to_cohere(cm: ConfigManager):
    """Configures the ConfigManager to use Cohere language and embedding models
    with the API key set. Returns a 3-tuple of the keyword arguments used."""
    LM_GID = "cohere:xlarge"
    EM_GID = "cohere:large"
    API_KEYS = {"COHERE_API_KEY": "foobar"}
    # the params use lowercase key names
    LM_LID = "xlarge"
    EM_LID = "large"
    API_PARAMS = {"cohere_api_key": "foobar"}
    req = UpdateConfigRequest(
        model_provider_id=LM_GID, embeddings_provider_id=EM_GID, api_keys=API_KEYS
    )
    cm.update_config(req)
    return LM_GID, EM_GID, LM_LID, EM_LID, API_PARAMS


def configure_to_openai(cm: ConfigManager):
    """Configures the ConfigManager to use OpenAI language and embedding models
    with the API key set. Returns a 3-tuple of the keyword arguments used."""
    LM_GID = "openai-chat-new:gpt-3.5-turbo"
    EM_GID = "openai:text-embedding-ada-002"
    API_KEYS = {"OPENAI_API_KEY": "foobar"}
    LM_LID = "gpt-3.5-turbo"
    EM_LID = "text-embedding-ada-002"
    API_PARAMS = {"openai_api_key": "foobar"}
    req = UpdateConfigRequest(
        model_provider_id=LM_GID, embeddings_provider_id=EM_GID, api_keys=API_KEYS
    )
    cm.update_config(req)
    return LM_GID, EM_GID, LM_LID, EM_LID, API_PARAMS


def test_snapshot_default_config(cm: ConfigManager, snapshot):
    config_from_cm: DescribeConfigResponse = cm.get_config()
    assert config_from_cm == snapshot(exclude=lambda prop, path: prop == "last_read")


def test_init_with_existing_config(cm: ConfigManager, common_cm_kwargs):
    configure_to_cohere(cm)
    del cm

    ConfigManager(**common_cm_kwargs)


def test_init_with_blocklists(cm: ConfigManager, common_cm_kwargs):
    configure_to_openai(cm)
    del cm

    blocked_providers = ["openai"]  # blocks EM
    blocked_models = ["openai-chat-new:gpt-3.5-turbo"]  # blocks LM
    kwargs = {
        **common_cm_kwargs,
        "blocked_providers": blocked_providers,
        "blocked_models": blocked_models,
    }
    test_cm = ConfigManager(**kwargs)
    assert test_cm._blocked_providers == blocked_providers
    assert test_cm._blocked_models == blocked_models
    assert test_cm.lm_gid == None
    assert test_cm.em_gid == None


def test_init_with_allowlists(cm: ConfigManager, common_cm_kwargs):
    configure_to_cohere(cm)
    del cm

    allowed_providers = ["openai"]  # blocks both LM & EM

    kwargs = {**common_cm_kwargs, "allowed_providers": allowed_providers}
    test_cm = ConfigManager(**kwargs)
    assert test_cm._allowed_providers == allowed_providers
    assert test_cm._allowed_models == None
    assert test_cm.lm_gid == None
    assert test_cm.em_gid == None


def test_property_access_on_default_config(cm: ConfigManager):
    """Asserts that the CM behaves well with an empty, default
    configuration."""
    assert cm.lm_gid == None
    assert cm.em_gid == None
    assert cm.lm_provider == None
    assert cm.lm_provider_params == None
    assert cm.em_provider == None
    assert cm.em_provider_params == None


def test_indentation_depth(common_cm_kwargs, config_path):
    """Asserts that the CM indents the configuration and respects the
    `indentation_depth` trait when specified."""
    INDENT_DEPTH = 7
    ConfigManager(**common_cm_kwargs, indentation_depth=INDENT_DEPTH)
    with open(config_path) as f:
        config_file = f.read()
        config_lines = config_file.split("\n")
        lm_gid_line = next(
            line for line in config_lines if '"model_provider_id":' in line
        )
        assert lm_gid_line.startswith(" " * INDENT_DEPTH)
        assert not lm_gid_line.startswith(" " * (INDENT_DEPTH + 1))


def test_describe(cm: ConfigManager):
    LM_GID, EM_GID, LM_LID, EM_LID, API_PARAMS = configure_to_cohere(cm)

    config_desc = cm.get_config()
    assert config_desc.model_provider_id == LM_GID
    assert config_desc.embeddings_provider_id == EM_GID
    assert config_desc.api_keys == ["COHERE_API_KEY"]
    assert cm.lm_provider_params == {**API_PARAMS, "model_id": LM_LID}
    assert cm.em_provider_params == {**API_PARAMS, "model_id": EM_LID}


def test_update(cm: ConfigManager):
    LM_GID, EM_GID, LM_LID, EM_LID, API_PARAMS = configure_to_cohere(cm)

    new_config = cm.get_config()
    assert new_config.model_provider_id == LM_GID
    assert new_config.embeddings_provider_id == EM_GID
    assert new_config.api_keys == ["COHERE_API_KEY"]
    assert cm.lm_provider_params == {**API_PARAMS, "model_id": LM_LID}
    assert cm.em_provider_params == {**API_PARAMS, "model_id": EM_LID}


def test_update_no_empty_field_dicts(cm: ConfigManager, config_path):
    LM_GID, _, _, _, _ = configure_to_cohere(cm)
    cm.update_config(UpdateConfigRequest(fields={LM_GID: {}}))

    with open(config_path) as f:
        raw_config = json.loads(f.read())
        assert raw_config["fields"] == {}


def test_update_fails_with_invalid_req():
    with pytest.raises(ValidationError):
        UpdateConfigRequest(send_with_shift_enter=None)
    with pytest.raises(ValidationError):
        UpdateConfigRequest(fields=None)
    with pytest.raises(ValidationError):
        UpdateConfigRequest(api_keys=None)


def test_update_fails_without_auth(cm: ConfigManager):
    LM_GID = "cohere:xlarge"
    EM_GID = "cohere:large"
    req = UpdateConfigRequest(model_provider_id=LM_GID, embeddings_provider_id=EM_GID)
    with pytest.raises(AuthError):
        cm.update_config(req)


def test_update_after_describe(cm: ConfigManager):
    configure_to_cohere(cm)
    last_read = cm.get_config().last_read
    cm.update_config(
        UpdateConfigRequest(model_provider_id="cohere:medium", last_read=last_read)
    )

    new_config = cm.get_config()
    assert new_config.model_provider_id == "cohere:medium"


# TODO: make the test work on Linux including CI.
@pytest.mark.skip(reason="Flakey on Linux including CI.")
def test_forbid_write_write_conflict(cm: ConfigManager):
    configure_to_openai(cm)
    # call DescribeConfig
    last_read = cm.get_config().last_read

    # call UpdateConfig separately after DescribeConfig with `last_read` unset
    # to force a write
    cm.update_config(UpdateConfigRequest(model_provider_id="openai-chat-new:gpt-4"))

    # this update should fail, as this generates a write-write conflict (where
    # the second update clobbers the first update).
    with pytest.raises(WriteConflictError):
        cm.update_config(
            UpdateConfigRequest(
                model_provider_id="openai-chat-new:gpt-4-32k", last_read=last_read
            )
        )


def test_update_api_key(cm: ConfigManager):
    """Asserts that updates via direct edits to the config file are immediately
    reflected by the ConfigManager."""
    LM_GID, EM_GID, LM_LID, EM_LID, _ = configure_to_cohere(cm)
    cm.update_config(UpdateConfigRequest(api_keys={"COHERE_API_KEY": "barfoo"}))

    config_desc = cm.get_config()
    assert config_desc.api_keys == ["COHERE_API_KEY"]
    assert cm.lm_provider_params == {"cohere_api_key": "barfoo", "model_id": LM_LID}
    assert cm.em_provider_params == {"cohere_api_key": "barfoo", "model_id": EM_LID}


def test_delete_api_key(cm: ConfigManager):
    configure_to_cohere(cm)
    cm.update_config(UpdateConfigRequest(api_keys={"OPENAI_API_KEY": "asdf"}))
    assert cm.get_config().api_keys == ["COHERE_API_KEY", "OPENAI_API_KEY"]

    cm.delete_api_key("OPENAI_API_KEY")
    assert cm.get_config().api_keys == ["COHERE_API_KEY"]


def test_forbid_deleting_key_in_use(cm: ConfigManager):
    configure_to_cohere(cm)

    with pytest.raises(KeyInUseError):
        cm.delete_api_key("COHERE_API_KEY")


def test_handle_bad_provider_ids(cm_with_bad_provider_ids):
    config_desc = cm_with_bad_provider_ids.get_config()
    assert config_desc.model_provider_id is None
    assert config_desc.embeddings_provider_id is None
