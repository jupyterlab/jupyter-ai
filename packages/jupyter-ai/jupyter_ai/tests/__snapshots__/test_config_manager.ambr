# serializer version: 1
# name: test_snapshot_default_config
  dict({
    'api_keys': list([
    ]),
    'completions_fields': dict({
    }),
    'completions_model_provider_id': None,
    'embeddings_provider_id': None,
    'fields': dict({
    }),
    'model_provider_id': None,
    'send_with_shift_enter': False,
  })
# ---
